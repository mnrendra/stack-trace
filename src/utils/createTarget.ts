import type { Target } from '../types'

/**
 * Create an initial stack target.
 *
 * @returns {Object} stack target.
 */
const createTarget = (): Target => ({
  stack: []
})

export default createTarget
