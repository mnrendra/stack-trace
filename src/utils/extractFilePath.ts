import { isAbsolute } from 'node:path'

import { normalizePath } from '../helpers'

/**
 * Extracts the file name from a `CallSite` object and converts it to a file
 * path if the value is a file URL.
 *
 * This utility ensures that the returned value is an absolute path.
 *
 * @param {NodeJS.CallSite} callSite - `CallSite` object captured from
 * {@link https://github.com/mnrendra/stack-trace#stacktrace `stackTrace`}.
 *
 * @returns {string} Absolute path of the file name extracted from a `CallSite`
 * object.
 *
 * @throws If the extracted file name is not a string or not
 * absolute.
 *
 * @see {@link https://github.com/mnrendra/stack-trace#extractfilepath documentation}
 */
const extractFilePath = (
  callSite: NodeJS.CallSite
): string => {
  const fileName = callSite.getFileName()

  const filePath = normalizePath(fileName)

  if (!isAbsolute(filePath)) {
    throw new Error(`Caller's file path is not absolute: "${filePath}"`)
  }

  return filePath
}

export default extractFilePath
