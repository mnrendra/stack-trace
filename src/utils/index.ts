import createTarget from './createTarget'

export {
  createTarget
}
