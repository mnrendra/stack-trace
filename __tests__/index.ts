import { testMethods } from '@tests/utils'

import { stackTrace } from '..'

describe('Test all APIs:', () => {
  describe('Test `stackTrace` API:', () => {
    it('Should capture the current stack trace when the callee is `undefined`!', () => {
      const caller = (): NodeJS.CallSite[] => stackTrace()
      const received = caller()
      expect(received).toMatchSnapshot()
      received.forEach(testMethods)
    })

    it('Should capture the current stack trace when the callee is `null`!', () => {
      const caller = (): NodeJS.CallSite[] => stackTrace(null)
      const received = caller()
      expect(received).toMatchSnapshot()
      received.forEach(testMethods)
    })

    it('Should capture the stack trace of the caller when the callee is a specific function!', () => {
      const callee = (): NodeJS.CallSite[] => stackTrace(callee)
      const caller = (): NodeJS.CallSite[] => callee()
      const received = caller()
      expect(received).toMatchSnapshot()
      received.forEach(testMethods)
    })

    it('Should only capture the first stack trace frame when the `limit` option is set to `1`!', () => {
      const caller = (): NodeJS.CallSite[] => stackTrace(caller, { limit: 1 })
      const received = caller()
      expect(received).toHaveLength(1)
      expect(received).toMatchSnapshot()
      received.forEach(testMethods)
    })

    describe('Test all methods of the first stack frame returned from the `stackTrace` API:', () => {
      const caller = (): NodeJS.CallSite[] => stackTrace()
      const [stack] = caller()

      it('Should return `null` when the `getTypeName` method is invoked!', () => {
        const received = stack.getTypeName()
        expect(received).toBeNull()
      })

      it('Should return the current `__filename` when the `getFileName` method is invoked!', () => {
        const received = stack.getFileName()
        const expected = __filename
        expect(received).toBe(expected)
      })

      it('Should return any `number` when the `getLineNumber` method is invoked!', () => {
        const received = stack.getLineNumber()
        const expected = expect.any(Number)
        expect(received).toEqual(expected)
      })

      it('Should return any `number` when the `getColumnNumber` method is invoked!', () => {
        const received = stack.getColumnNumber()
        const expected = expect.any(Number)
        expect(received).toEqual(expected)
      })

      it('Should return `null` when the `getMethodName` method is invoked!', () => {
        const received = stack.getMethodName()
        expect(received).toBeNull()
      })

      it('Should return the caller name when the `getFunctionName` method is invoked!', () => {
        const received = stack.getFunctionName()
        const expected = 'caller'
        expect(received).toBe(expected)
      })

      it('Should return `undefined` when the `getFunction` method is invoked!', () => {
        const received = stack.getFunction()
        expect(received).toBeUndefined()
      })

      it('Should return `undefined` when the `getThis` method is invoked!', () => {
        const received = stack.getThis()
        expect(received).toBeUndefined()
      })

      it('Should return `undefined` when the `getEvalOrigin` method is invoked!', () => {
        const received = stack.getEvalOrigin()
        expect(received).toBeUndefined()
      })

      it('Should return `false` when the `isConstructor` method is invoked!', () => {
        const received = stack.isConstructor()
        const expected = false
        expect(received).toBe(expected)
      })

      it('Should return `false` when the `isEval` method is invoked!', () => {
        const received = stack.isEval()
        const expected = false
        expect(received).toBe(expected)
      })

      it('Should return `false` when the `isNative` method is invoked!', () => {
        const received = stack.isNative()
        const expected = false
        expect(received).toBe(expected)
      })

      it('Should return `true` when the `isToplevel` method is invoked!', () => {
        const received = stack.isToplevel()
        const expected = true
        expect(received).toBe(expected)
      })
    })
  })
})
