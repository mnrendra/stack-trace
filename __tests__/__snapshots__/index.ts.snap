// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test all APIs: Test \`stackTrace\` API: Should capture the current stack trace when the callee is \`null\`! 1`] = `
[
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
]
`;

exports[`Test all APIs: Test \`stackTrace\` API: Should capture the current stack trace when the callee is \`undefined\`! 1`] = `
[
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
]
`;

exports[`Test all APIs: Test \`stackTrace\` API: Should capture the stack trace of the caller when the callee is a specific function! 1`] = `
[
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
  CallSite {},
]
`;

exports[`Test all APIs: Test \`stackTrace\` API: Should only capture the first stack trace frame when the \`limit\` option is set to \`1\`! 1`] = `
[
  CallSite {},
]
`;
